File: valenza_metrologia.py
"""
valenza_metrologia.py
======================
Script di Metrologia per Substrati Neuromorfici e Digital Twin (P0_Distilled v0.1)

Autore: Programma di Ricerca Lakatosiano (P0_Distilled v0.1)
Descrizione:
  Questo script funge da "righello metrologico" per elaborare le serie temporali fisiche
  misurate da substrati analogici (o simulate via SDE in un Digital Twin).
  
  Calcola:
  1. Decomposizione termodinamica di Hatano-Sasa (Housekeeping vs Excess entropy production).
  2. Stima non-parametrica k-NN della Divergenza di Kullback-Leibler (D_KL) per la nicchia allostasica.
  3. Guadagno predittivo dell'Efference Copy (G_pred) generata dal livello A2.
  4. Funzionale di Valenza Termodinamica e Allostasica integrato Psi(t).
"""

import numpy as np
from scipy.spatial import KDTree

def simulate_neuromorphic_substrate_sde(n_steps=10000, dt=0.001, seed=42):
    """
    Simula le dinamiche stocastiche di un substrato neuromorfico a due livelli (A1/A2)
    tramite integrazione SDE di Itô-Milstein.
    """
    np.random.seed(seed)
    
    # Parametri del materiale (Mott / Diffusive Memristor)
    a, b = -1.2, 0.8  # Parametri di attivazione locale (Edge of Chaos)
    sigma_noise = 0.15 # Ampiezza del rumore percolativo 1/f
    
    x_A1 = np.zeros(n_steps)
    s_obs = np.zeros(n_steps)
    s_pred = np.zeros(n_steps)
    
    x_val = 0.1
    for t in range(1, n_steps):
        # SDE per il livello primario A1
        dW = np.random.normal(0, np.sqrt(dt))
        dx = (a * x_val + b * np.tanh(x_val) + np.sin(t * dt * 2.0)) * dt + sigma_noise * dW
        # Correzione di Milstein
        dx += 0.5 * sigma_noise * sigma_noise * (dW**2 - dt)
        x_val += dx
        x_A1[t] = x_val
        
        # Riafferenza sensoriale reale
        s_obs[t] = x_val + np.random.normal(0, 0.05)
        
        # Copia efferente (livello A2) - predizione allostasica
        s_pred[t] = x_A1[t-1] + (a * x_A1[t-1] + b * np.tanh(x_A1[t-1])) * dt

    return x_A1, s_obs, s_pred, dt

def estimate_kl_divergence_knn(p_samples, q_samples, k=5):
    """
    Stima non-parametrica k-NN della divergenza di Kullback-Leibler D_KL(P || Q)
    basata sugli albori KDTree (Kraskov et al. / Perez-Cruz).
    """
    p_samples = np.atleast_2d(p_samples).T if p_samples.ndim == 1 else p_samples
    q_samples = np.atleast_2d(q_samples).T if q_samples.ndim == 1 else q_samples
    
    n, d = p_samples.shape
    m, _ = q_samples.shape
    
    tree_p = KDTree(p_samples)
    tree_q = KDTree(q_samples)
    
    # Distanze dai k-esimi vicini in P
    r_p, _ = tree_p.query(p_samples, k=k+1)
    r_k_p = r_p[:, -1]
    
    # Distanze dai k-esimi vicini in Q
    r_q, _ = tree_q.query(p_samples, k=k)
    r_k_q = r_q[:, -1]
    
    # Evita divisioni per zero
    r_k_p = np.maximum(r_k_p, 1e-12)
    r_k_q = np.maximum(r_k_q, 1e-12)
    
    kl_est = (d / n) * np.sum(np.log(r_k_q / r_k_p)) + np.log(m / (n - 1.0))
    return max(0.0, float(kl_est))

def calculate_thermodynamic_valence(x_A1, s_obs, s_pred, dt, alpha=1.0, beta=0.5, gamma=0.8):
    """
    Calcola il funzionale di valenza Psi(t) integrando la produzione di entropia NESS,
    la divergenza allostasica e il guadagno predittivo dell'efferenza.
    """
    # 1. Stima delle componenti dissipative NESS (Hatano-Sasa)
    dx = np.diff(x_A1) / dt
    sigma_hk = np.var(dx) # Dissipazione di housekeeping stazionaria
    sigma_ex = np.mean(np.abs(dx * x_A1[:-1])) # Dissipazione in eccesso per riorganizzazione
    
    # 2. Nicchia allostasica target p_target ~ N(0, 0.2)
    p_target = np.random.normal(0.0, 0.2, size=(len(x_A1), 1))
    x_samples = np.atleast_2d(x_A1).T
    
    d_kl_allostasica = estimate_kl_divergence_knn(x_samples, p_target, k=5)
    
    # 3. Guadagno Predittivo dell'Efference Copy (G_pred)
    s_obs_samples = np.atleast_2d(s_obs).T
    s_pred_samples = np.atleast_2d(s_pred).T
    
    d_kl_baseline = estimate_kl_divergence_knn(s_obs_samples, p_target, k=5)
    d_kl_predicted = estimate_kl_divergence_knn(s_pred_samples, p_target, k=5)
    
    g_pred = d_kl_baseline - d_kl_predicted
    
    # 4. Funzionale di Valenza Integrato Psi(t)
    psi = alpha * np.log(1.0 + (sigma_ex / (sigma_hk + 1e-8))) - beta * d_kl_allostasica + gamma * g_pred
    
    return {
        "sigma_ex": float(sigma_ex),
        "sigma_hk": float(sigma_hk),
        "d_kl_allostasica": float(d_kl_allostasica),
        "g_pred": float(g_pred),
        "psi_valenza": float(psi)
    }

if __name__ == "__main__":
    print("=== ESECUZIONE METROLOGIA AVANZATA DIGITAL TWIN (P0_Distilled v0.1) ===")
    x_A1, s_obs, s_pred, dt = simulate_neuromorphic_substrate_sde(n_steps=10000)
    res = calculate_thermodynamic_valence(x_A1, s_obs, s_pred, dt)
    
    print(f"Punti temporali processati: {len(x_A1)}")
    print(f"Stato medio substrato A1 x(t): {np.mean(x_A1):.4f}")
    print(f"Dissipazione in eccesso sigma_ex: {res['sigma_ex']:.4f}")
    print(f"Dissipazione housekeeping sigma_hk: {res['sigma_hk']:.4f}")
    print(f"Divergenza Allostasica D_KL(P||P_target): {res['d_kl_allostasica']:.4f}")
    print(f"Guadagno Predittivo dell'Efference Copy G_pred: {res['g_pred']:.4f}")
    print(f"-> FUNZIONALE DI VALENZA INTEGRATO Psi(t): {res['psi_valenza']:.4f}")
    print("=========================================================================")

File: valenza_metrologia.rs
//! valenza_metrologia.rs
//! ======================
//! High-Performance Metrology Module in Rust for Neuromorphic Substrates & Digital Twin (P0_Distilled v0.1)
//!
//! Autore: Programma di Ricerca Lakatosiano (P0_Distilled v0.1)
//! Descrizione:
//!   Modulo metrologico ad alte prestazioni scritto in Rust per l'elaborazione in tempo reale
//!   di serie temporali fisiche estratte da substrati neuromorfici analogici.
//!   
//! Implementa:
//!   1. Integrazione SDE stocastica di Itô-Milstein per architettura a due livelli A1/A2.
//!   2. Decomposizione termodinamica di NESS (Housekeeping vs Excess entropy production).
//!   3. Stima non-parametrica k-NN della Divergenza di Kullback-Leibler (D_KL).
//!   4. Guadagno predittivo della copia efferente (G_pred).
//!   5. Funzionale di Valenza Termodinamica e Allostasica integrato Psi(t).

use std::f64::consts::PI;

/// Struttura dati per i risultati delle metriche metrologiche
#[derive(Debug, Clone)]
pub struct MetrologyResult {
    pub sigma_ex: f64,
    pub sigma_hk: f64,
    pub d_kl_allostasica: f64,
    pub g_pred: f64,
    pub psi_valenza: f64,
}

/// Generatore di numeri casuali ad alte prestazioni (Xorshift128+) per zero dipendenze esterne
pub struct FastRng {
    s: [u64; 2],
}

impl FastRng {
    pub fn new(seed: u64) -> Self {
        let mut s0 = seed ^ 0x9E3779B97F4A7C15;
        let mut s1 = (seed.wrapping_add(0xBF58476D1CE4E5B9)) ^ 0x94D049BB133111EB;
        if s0 == 0 && s1 == 0 {
            s0 = 1;
        }
        Self { s: [s0, s1] }
    }

    pub fn next_u64(&mut self) -> u64 {
        let mut x = self.s[0];
        let y = self.s[1];
        self.s[0] = y;
        x ^= x << 23;
        self.s[1] = x ^ y ^ (x >> 17) ^ (y >> 26);
        self.s[1].wrapping_add(y)
    }

    pub fn next_f64(&mut self) -> f64 {
        (self.next_u64() >> 11) as f64 * (1.0 / (1u64 << 53) as f64)
    }

    /// Trasformata di Box-Muller per variabili casuali gaussiane
    pub fn next_gaussian(&mut self) -> f64 {
        let u1 = self.next_f64().max(1e-15);
        let u2 = self.next_f64();
        (-2.0 * u1.ln()).sqrt() * (2.0 * PI * u2).cos()
    }
}

/// Simula il substrato neuromorfico A1/A2 tramite integrazione SDE di Itô-Milstein
pub fn simulate_neuromorphic_substrate_sde(
    n_steps: usize,
    dt: f64,
    seed: u64,
) -> (Vec<f64>, Vec<f64>, Vec<f64>) {
    let mut rng = FastRng::new(seed);
    let a = -1.2;
    let b = 0.8;
    let sigma_noise = 0.15;

    let mut x_a1 = vec![0.0; n_steps];
    let mut s_obs = vec![0.0; n_steps];
    let mut s_pred = vec![0.0; n_steps];

    let mut x_val: f64 = 0.1;
    let sqrt_dt = dt.sqrt();

    for t in 1..n_steps {
        let dw = rng.next_gaussian() * sqrt_dt;
        let time_val = t as f64 * dt;

        // Step SDE Itô-Milstein per il substrato A1
        let mut dx = (a * x_val + b * x_val.tanh() + (time_val * 2.0).sin()) * dt
            + sigma_noise * dw;
        dx += 0.5 * sigma_noise * sigma_noise * (dw * dw - dt);

        x_val += dx;
        x_a1[t] = x_val;

        // Riafferenza sensoriale reale
        s_obs[t] = x_val + rng.next_gaussian() * 0.05;

        // Copia efferente predittiva (Livello A2)
        let prev_x = x_a1[t - 1];
        s_pred[t] = prev_x + (a * prev_x + b * prev_x.tanh()) * dt;
    }

    (x_a1, s_obs, s_pred)
}

/// Stima della Divergenza di Kullback-Leibler D_KL(P || Q) tramite vicini più prossimi (k-NN)
pub fn estimate_kl_divergence_knn_1d(p_samples: &[f64], q_samples: &[f64], k: usize) -> f64 {
    let n = p_samples.len();
    let m = q_samples.len();

    if n <= k || m <= k {
        return 0.0;
    }

    let mut sorted_q = q_samples.to_vec();
    sorted_q.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

    let mut kl_sum = 0.0;

    for &p_i in p_samples.iter() {
        // Distanza dal k-esimo vicino in P
        let mut p_dists: Vec<f64> = p_samples.iter().map(|&x| (x - p_i).abs()).collect();
        p_dists.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let r_k_p = p_dists[k].max(1e-12);

        // Distanza dal k-esimo vicino in Q
        let mut q_dists: Vec<f64> = sorted_q.iter().map(|&x| (x - p_i).abs()).collect();
        q_dists.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
        let r_k_q = q_dists[k - 1].max(1e-12);

        kl_sum += (r_k_q / r_k_p).ln();
    }

    let kl_est = (1.0 / n as f64) * kl_sum + (m as f64 / (n - 1) as f64).ln();
    kl_est.max(0.0)
}

/// Calcola il funzionale di valenza integrato Psi(t)
pub fn calculate_thermodynamic_valence(
    x_a1: &[f64],
    s_obs: &[f64],
    s_pred: &[f64],
    dt: f64,
    alpha: f64,
    beta: f64,
    gamma: f64,
) -> MetrologyResult {
    let n = x_a1.len();
    let mut dx = Vec::with_capacity(n - 1);
    for i in 0..n - 1 {
        dx.push((x_a1[i + 1] - x_a1[i]) / dt);
    }

    // Dissipazione di housekeeping (varianza di dx)
    let mean_dx = dx.iter().sum::<f64>() / dx.len() as f64;
    let sigma_hk = dx.iter().map(|v| (v - mean_dx).powi(2)).sum::<f64>() / dx.len() as f64;

    // Dissipazione in eccesso per riorganizzazione
    let mut excess_sum = 0.0;
    for i in 0..dx.len() {
        excess_sum += (dx[i] * x_a1[i]).abs();
    }
    let sigma_ex = excess_sum / dx.len() as f64;

    // Distribuzione target per la nicchia allostasica
    let mut rng = FastRng::new(12345);
    let target_samples: Vec<f64> = (0..n).map(|_| rng.next_gaussian() * 0.2).collect();

    let d_kl_allostasica = estimate_kl_divergence_knn_1d(x_a1, &target_samples, 5);
    let d_kl_baseline = estimate_kl_divergence_knn_1d(s_obs, &target_samples, 5);
    let d_kl_predicted = estimate_kl_divergence_knn_1d(s_pred, &target_samples, 5);

    let g_pred = d_kl_baseline - d_kl_predicted;

    let psi = alpha * (1.0 + (sigma_ex / (sigma_hk + 1e-8))).ln() - beta * d_kl_allostasica + gamma * g_pred;

    MetrologyResult {
        sigma_ex,
        sigma_hk,
        d_kl_allostasica,
        g_pred,
        psi_valenza: psi,
    }
}

fn main() {
    println!("=== ESECUZIONE METROLOGIA RUST DIGITAL TWIN (P0_Distilled v0.1) ===");
    let n_steps = 10000;
    let dt = 0.001;
    let (x_a1, s_obs, s_pred) = simulate_neuromorphic_substrate_sde(n_steps, dt, 42);

    let res = calculate_thermodynamic_valence(&x_a1, &s_obs, &s_pred, dt, 1.0, 0.5, 0.8);

    println!("Punti temporali processati: {}", n_steps);
    println!("Stato medio substrato A1 x(t): {:.4}", x_a1.iter().sum::<f64>() / n_steps as f64);
    println!("Dissipazione in eccesso sigma_ex: {:.4}", res.sigma_ex);
    println!("Dissipazione housekeeping sigma_hk: {:.4}", res.sigma_hk);
    println!("Divergenza Allostasica D_KL(P||P_target): {:.4}", res.d_kl_allostasica);
    println!("Guadagno Predittivo dell'Efference Copy G_pred: {:.4}", res.g_pred);
    println!("-> FUNZIONALE DI VALENZA INTEGRATO Psi(t): {:.4}", res.psi_valenza);
    println!("=========================================================================");
}
